droidlet/dashboard/web/src/components/Interact/Question.js

```diff
@@ -337,36 +277,288 @@ class Question extends Component {
+        }
+        let antecedent = [qty, "", "", "", ""]; // qty then size then colour then block type then name. Ignore everything else.
+        action_dict.reference_object.filters.where_clause.AND.forEach(
+          (clause) => {
+            if (clause.pred_text === "has_size")
+              antecedent[1] = clause.obj_text;
+            else if (clause.pred_text === "has_colour")
+              antecedent[2] = clause.obj_text;
+            else if (clause.pred_text === "has_block_type")
+              antecedent[3] = clause.obj_text;
+            else if (clause.pred_text === "has_name")
+              antecedent[4] = clause.obj_text;
+          }
+        );
+        locationRef =
+          "'" + antecedent.join(" ").replace(/  +/g, " ").trim() + "'";
+      }
+    }
+    return locationRef;
+  }
+
+  evalCommandPerception() {
+    //                 Reference Object Exists in dictionary ?
+    //                         /   \
+    //                       Yes    No -> Other Error
+    //                        |
+    //             Found Location in Memory?
+    //                      /   \
+    //                    Yes    No -> Perception Error
+    let ref_object = false;
+    let reference_object_description = null;
+    // Check if reference object exists in the dictionary anywhere
+    if (this.state.action_dict) {
+      if (this.state.action_dict["dialogue_type"] == "HUMAN_GIVE_COMMAND") {
+        // also implement for get and put memory
+        for (const action of this.state.action_dict.action_sequence) {
+          ref_object = this.check_reference_object_in_action_dict(action);
+        }
+      }
+
+      // If yes, find reference object description.
+      if (ref_object == true) {
+        const action_dict = this.state.action_dict.action_sequence[0];
+        // Check for location at top level and extract the reference text
+        let considered_action_dict = null;
+        if ("location" in action_dict) {
+          considered_action_dict = action_dict.location;
+        } else if ("reference_object" in action_dict) {
+          considered_action_dict = action_dict;
+        }
+
+        if (considered_action_dict) {
+          reference_object_description = this.extractLocationRef(
+            considered_action_dict
+          );
+        }
+      }
+      // If no reference object description found no perception error.
+    } else {
+      console.log("no action dictionary found ..."); // Shouldn't happen....
+    }
+    this.setState({
+      reference_object_description: reference_object_description,
+    });
+
+    // else {
+    //   // shouldn't happen
+    //   return (
+    //     <div>
+    //       <h3> Thanks! Press to continue.</h3>
+    //       <Button
+    //         variant="contained"
+    //         color="primary"
+    //         onClick={() => this.props.goToMessage()}
+    //       >
+    //         Done
+    //       </Button>
+    //     </div>
+    //   );
+    // }
```

Comment:
Probably don't need to preserve this, unless we think we want a different UX than people being shown a blank description in the error state.

Reply:
Same as above ^
This wasn't documented so I wasn't sure if I can just torch, can torch this